text/0000-third-party-web-components-support.md

```diff
@@ -0,0 +1,103 @@
+While there are several ways to do this today (e.g., through the use of iframes or the
+`lwc:dom="manual"` directive), the preferred way would be to enable this use case through the
+rendering of native web components. Browser support for native web components was still in the early
+stages when LWC was first implemented, but it is quite good today. In addition, many frameworks now
+support native web components as a compilation target so it makes sense to add native web component
+support as a first class citizen.
+
+## Detailed design
+
+This section covers the identified work to add native web component support.
+
+### Constructor resolution
+
+By convention, LWC implicitly resolves component constructors for every custom element in its
+template. This behavior needs to be modified to allow authors the flexibility to resolve their own
+components.
+
+#### Proposed solution
+
+LWC would provide a directive (e.g., `lwc:external`) which signals that component resolution should
+be delegated to the browser. LWC will not attempt to resolve components for custom elements tagged
+with this directive. If it happens that the component was not registered, the custom element will
+simply render as an HTMLUnknownElement, which is the default browser behavior.
+
+### Passing data
+
+By convention, custom elements are assumed to be backed by LWC components and values are set through
+their properties. Given that a native web component might expect data passed as an attribute or as a
+property, we would need to support both mechanisms. There are multiple approaches we could take
+here, and the following options are how some popular frameworks do it today. We could default to
+either attributes or properties and introduce a directive that would act as a signal to the engine
+to use one or the other, we could implement a runtime heuristic to choose one over the other, or we
+could default to attributes and set properties when setting a data structure.
+
+#### Proposed solution
+
+When passing data to a native web component, we can take a heuristical approach where if the
+property exists, the property is set. Otherwise the attribute is set. While it may be advantageous
+to require an explicit signal in the form of a directive, this approach seems to strike a good
+balance between developer experience and API semantics.
+
+### Events
+
+EventTarget polyfilled methods and properties such as `add/removeEventListener` and `target` go
+through the synthetic shadow polyfill due to those methods and properties being globally patched.
+This is the current status quo for native shadow roots and should not be an issue for supporting
+third party web components.
+
+LWC's declarative event bindings only support lowercase events. Events using anything other than
+lowercase must be listened to imperatively. This would not need to change in order to support native
+web components.

```

Comment:
The test suite for [Custom Elements Everywhere](https://custom-elements-everywhere.com/), which is basically the industry yardstick for "does a framework support web components," explicitly checks to see if events with `names-like-this` or `namesLikeThis` can be listened for in a declarative way. (They don't consider it good enough that you can always call `addEventListener`.)

I think maybe we should consider relaxing our requirement that every event is lowercase without hyphens, since otherwise we won't score 100% on this test.

Reply:
I added some more context around event names in this rfc. Will schedule the topic in a dev sync and mark this as resolved.